bundles/org.openhab.binding.gree/src/main/java/org/openhab/binding/gree/internal/handler/GreeAirDevice.java

```diff
@@ -0,0 +1,713 @@
+            String modifiedSentence = new String(receivePacket.getData(), "UTF-8");
+
+            // Keep a copy of the old response to be used to check if values have changed
+            // If first time running, there will not be a previous GreeStatusResponsePack4Gson
+            if (statusResponseGson != null && statusResponseGson.packJson != null) {
+                prevStatusResponsePackGson = new GreeStatusResponsePack4GsonDTO(statusResponseGson.packJson);
+            }
+
+            // Read the response
+            StringReader stringReader = new StringReader(modifiedSentence);
+            statusResponseGson = gson.fromJson(new JsonReader(stringReader), GreeStatusResponse4GsonDTO.class);
+            statusResponseGson.decryptedPack = GreeCryptoUtil.decryptPack(this.getKey().getBytes(),
+                    statusResponseGson.pack);
+
+            logger.trace("Response from device: {}", statusResponseGson.decryptedPack);
+
+            // Create the JSON to hold the response values
+            stringReader = new StringReader(statusResponseGson.decryptedPack);
+
+            statusResponseGson.packJson = gson.fromJson(new JsonReader(stringReader),
+                    GreeStatusResponsePack4GsonDTO.class);
+            updateTempFtoC();
+        } catch (IOException e) {
+            throw new GreeException(e, "I/O exception while receiving data");
+        }
+    }
+
+    private void updateTempFtoC() {
+        // Status message back from A/C always reports degrees C
+        // If using Fahrenheit, us SetTem, TemUn and TemRec to reconstruct the Fahrenheit temperature
+        // Get Celsius or Fahrenheit from status message
+        int CorF = getIntStatusVal("TemUn");
+        int newVal = getIntStatusVal("SetTem");
+        int halfStep = getIntStatusVal("TemRec");
+
+        if ((CorF == -1) || (newVal == -1) || (halfStep == -1)) {
+            throw new IllegalArgumentException("SetTem,TemUn or TemRec is invalid, not performing conversion");
+        } else if (CorF == 1) { // convert SetTem to Fahrenheit
+            // Find the valueName in the Returned Status object
+            String columns[] = statusResponseGson.packJson.cols;
+            Integer values[] = statusResponseGson.packJson.dat;
+            List<String> colList = new ArrayList<>(Arrays.asList(columns));
+            int valueArrayposition = colList.indexOf("SetTem");
+            if (valueArrayposition != -1) {
+                // convert Celsius to Fahrenheit,
+                // SetTem status returns degrees C regardless of TempUn setting
+
+                // Perform the float Celsius to Fahrenheit conversion add or subtract 0.5 based on the value of TemRec
+                // (0 = -0.5, 1 = +0.5). Pass into a rounding function, this yeild the correct Fahrenheit Temperature to
+                // match A/C display
+                newVal = (int) (Math.round(((newVal * 9.0 / 5.0) + 32.0) + halfStep - 0.5));
```

Comment:
that's really overdriven, it's simple math with primitives

Reply:
Simple math, yes, but using the units framework makes it much more obvious what your intent is without risk of a computation mistake, thus making the code more readable and eliminates the risks of a simple typo (like using int instead of doubles) from breaking things.

Reply:
ok, I refactored temp handling completely. Now conversion is based on UBM and the binding maps the requested unit to the AC command

Reply:
done